src/Catty/ViewController/Upload/UploadViewController.swift

```diff
@@ -259,171 +203,42 @@ class UploadViewController: UIViewController {
-                        if let aTag = dictionary?[self.tokenParameterTag] {
-                            JNKeychain.saveValue(aTag, forKey: kUserLoginToken)
-                        }
-
-                        DispatchQueue.main.async(execute: {
-                            self.dismissView()
-                        })
-                    } else {
-                        let serverResponse = dictionary?[self.answerTag] as? String
-                        debugPrint("Error: ", serverResponse ?? "undefined")
-                        DispatchQueue.main.async(execute: {
-                            if serverResponse != nil {
-                                Util.alert(withText: serverResponse)
-                            } else {
-                                Util.alert(withText: kLocalizedUploadProblem)
-                            }
-                        })
-
-                        if statusCode == self.statusCodeTokenWrong {
-                            //Token not valid
-                            UserDefaults.standard.set(false, forKey: kUserIsLoggedIn)
-
-                            var viewArray: [AnyHashable]?
-                            if let aControllers = self.parent?.navigationController?.viewControllers {
-                                viewArray = aControllers
-                            }
-                            viewArray?.removeLast()
-                            var newViewArray: [AnyHashable]?
-                            if let anArray = viewArray {
-                                newViewArray = anArray
-                            }
-                            if let anArray = newViewArray as? [UIViewController] {
-                                self.parent?.navigationController?.setViewControllers(anArray, animated: true)
-                            }
-                        }
-                    }
-                }
+        if let project = project {
+            if let uploader = self.uploader {
+                uploader.upload(project: project,
+                                completion: { error in
+                                    self.enableUploadView()
+                                    if let error = error {
+                                        switch error {
+                                        case .request(let error, _):
+                                            Util.alert(withText: error?.localizedDescription)
+                                        case .unexpectedError:
+                                            Util.defaultAlertForNetworkError()
+                                        case .zippingError, .timeout:
+                                            Util.alert(withText: kLocalizedUploadProblem)
+                                        case .invalidProject, .authenticationFailed:
+                                            //Token or UserName not valid
+                                            UserDefaults.standard.set(false, forKey: kUserIsLoggedIn)
```

Comment:
In case of an authentication error display the alert "Session expired, please log in." (add _kLocalizedSessionExpired_ to _LanguageTranslationDefines.h_) and redirect to the `LoginViewController`.